### PR TITLE
[US-3861] Monorepo split allow adding new package when monorepo is already tagged

### PR DESCRIPTION
**Pull request by simara-svatopluk**
| Q             | A
| ------------- | ---
|Description, reason for the PR| US-3861
|New feature| Yes
|BC breaks| No
|Fixes issues| ...
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

**Comment by Miroslav-Stopka** [comment]
Hi @simara-svatopluk,

thank you for your contribution :) Im ready for review and tests.

File: packages/monorepo-tools/rewrite_history_from.sh [hunk 1]
```diff
@@ -10,8 +10,25 @@
...
-REV_LIST_PARAMS=${@:2}
+REV_LIST_PARAMS=`echo ${@:2} | tr " " "\n"`
 echo "Rewriting history from a subdirectory '$SUBDIRECTORY'"
+
+# Get directory of the other scripts
+MONOREPO_SCRIPT_DIR=$(dirname "$0")
+# Backup tag refs because `git filter-branch` doesn't backup tags
+$MONOREPO_SCRIPT_DIR/tag_refs_backup.sh
+
+# Remove tags that don't contain any files in $SUBDIRECTORY
```
Comment (Miroslav-Stopka) [inline packages/monorepo-tools/rewrite_history_from.sh]: Im am kind of dumb user Svatopluk. So I like clear and describing comments. This comment means that tag can contain some files?

File: packages/monorepo-tools/rewrite_history_from.sh [hunk 2]
```diff
@@ -10,8 +10,25 @@
...
 
 SUBDIRECTORY=$1
-REV_LIST_PARAMS=${@:2}
+REV_LIST_PARAMS=`echo ${@:2} | tr " " "\n"`
 echo "Rewriting history from a subdirectory '$SUBDIRECTORY'"
+
+# Get directory of the other scripts
+MONOREPO_SCRIPT_DIR=$(dirname "$0")
+# Backup tag refs because `git filter-branch` doesn't backup tags
+$MONOREPO_SCRIPT_DIR/tag_refs_backup.sh
```
Comment (Miroslav-Stopka) [inline packages/monorepo-tools/rewrite_history_from.sh]: First you backup some tags. Next, you remove some tags. Why?

File: packages/monorepo-tools/rewrite_history_from.sh [hunk 3]
```diff
@@ -35,3 +52,5 @@ SUBDIRECTORY=$SUBDIRECTORY SUBDIRECTORY_SED=${SUBDIRECTORY//-/\\-} TAB=$'\t' XAR
     --tag-name-filter 'cat' \
     -- $REV_LIST_PARAMS
 
+# Moves backup tag refs to the same group with automatic backup from `git filter-branch`
```
Comment (Miroslav-Stopka) [inline packages/monorepo-tools/rewrite_history_from.sh]: So you backup all tags. Next, you are removing some tags - tags that are unnecessary for an actual split of a package, finally, you restore all tags in monorepo. Am I right?

File: packages/monorepo-tools/rewrite_history_into.sh [hunk 4]
```diff
@@ -12,6 +12,12 @@
 SUBDIRECTORY=$1
 REV_LIST_PARAMS=${@:2}
 echo "Rewriting history into a subdirectory '$SUBDIRECTORY'"
+
```
Comment (Miroslav-Stopka) [inline packages/monorepo-tools/rewrite_history_into.sh]: Was this change necessary? rewrite_history_into.sh is used only during the build of monorepo. Why you backup and restore tags during the build?

File: packages/monorepo-tools/README.md [hunk 5]
```diff
@@ -123,6 +123,18 @@ Delete all local branches and create all non-remote-tracking branches of a speci
```
Comment (Miroslav-Stopka) [inline packages/monorepo-tools/README.md]: We have new function :) New package is allowed by split of monorepo. I would like to see a brief tutorial on how to split the new package. I need only two or three sentences in the new section before the Reference section

**Comment by Miroslav-Stopka** [comment]
@simara-svatopluk I have already tested this change (it was not an easy one, because I have tried to make it functional with git version 2.10 :D ) But in the end, everything worked fine. I have just some questions. Please, answer them :)

File: packages/monorepo-tools/rewrite_history_from.sh (hunk 3 above)
Comment (simara-svatopluk) [reply packages/monorepo-tools/rewrite_history_from.sh]: Almost exactly!
1. Backup monorepo tags :heavy_check_mark: 
2. Remove tags that are unnecessary for split :heavy_check_mark: 
3. Restoring is not the right word, because monorepo tags are not restored, they are "restored"/moved into refs/original section, so they can be restored by git later.

File: packages/monorepo-tools/rewrite_history_into.sh (hunk 4 above)
Comment (simara-svatopluk) [reply packages/monorepo-tools/rewrite_history_into.sh]: We agreed on this during PIA.
Since git do not backup tags during `git filter-branch`, we decided to do so even during creating monorepo.
Behavior before this change is strange - git does `filter-branch`, moves tags. Then git is reseted to original state, but tags are still on commits after rewriting history into. Then second package is processed, but tags still aims to first package commits. And in the end `filter-branch` deals with this strange situation "somehow".
This change makes this strange situation more straightforward - backup all tags, filter-branch, restore all tags.

File: packages/monorepo-tools/rewrite_history_from.sh (hunk 1 above)
Comment (simara-svatopluk) [reply packages/monorepo-tools/rewrite_history_from.sh]: contains means affected

File: packages/monorepo-tools/rewrite_history_into.sh (hunk 4 above)
Comment (simara-svatopluk) [reply packages/monorepo-tools/rewrite_history_into.sh]: this comment is invalid. Change reverted from rewrite_history_into.sh
